docs/authentication.md

```diff
@@ -0,0 +1,142 @@
+Authentication
+==============
+
+Charmed Kubeflow uses [Dex][dex] for authentication. Dex allows for pluggable authentication
+against many different identity providers, while presenting a unified
+[OpenId Connect (OIDC)][oidc] interface.
+
+[dex]: https://github.com/dexidp/dex
+[oidc]: https://en.wikipedia.org/wiki/OpenID_Connect
+
+
+Auth Flow
+---------
+
+The authentication flow for an unauthenticated request consists of these steps:
+
+ - User makes an unauthenticated request
+   - Redirected to /dex
+ - User logs in via Dex
+ - Redirected to gatekeeper callback endpoint
+   - Sets a cookie with auth token for future requests
+ - Redirected back to original page, with authorization token
+
+### Unauthenticated Request
+
+The flow of an unauthenticated request is shown here:
+
+<img
+    src="dex-unauthenticated.svg"
+    alt="Unauthenticated Request"
+    height="500" />
+
+ - A user makes a request that goes through Ambassador
+ - Ambassador checks with the gatekeeper service before allowing any request to go through
+ - The gatekeeper service responds to Ambassador that the request is unauthenticated, and a redirect URL
+ - Ambassador returns an HTTP 301 redirecting user to Dex
+
+### Logging In
+
+<img
+    src="img/dex-login.svg"
+    alt="Logging In"
+    height="500" />
+
+ - User makes request to /dex
+ - Ambassador is configured to not check with the gatekeeper for requests to /dex
+ - Dex presents a login page to the user
+ - The user submits their credentials
+ - Dex uses the configured connector to authenticate the user against an external auth service
+   - Dex may also be configured with basic username/password support. This is the default in Charmed Kubeflow.
+   - See the [dex configuration](#configuring-dex) section for more information on how to configure Dex with
+     other connectors.
+ - Dex redirects the user to a callback URL managed by the gatekeeper
+
+
+### Receive Token
+
+<img
+    src="img/dex-callback.svg"
+    alt="Receive Token"
+    height="500" />
+
+ - User makes request to callback URL
+ - Ambassador sends request to gatekeeper service
+ - Gatekeeper service generates JWT token in the `Set-Cookie` response header
+
+### Authenticated Request
+
```

Comment:
I suspect we are better off putting images in the asset server, but that can be done later.